tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/ControlledWorkerScheduler.java

```diff
@@ -201,32 +201,27 @@ public void toss(Exception e) {
 	class WorkerRunnable implements Runnable {
 
 		private final ParallelTask<T> task;
+		private final ParallelExecutor<T> taskControl;
 
 		private volatile boolean aborted = false;
 
 		public WorkerRunnable(ParallelTask<T> task) {
 			super();
 			this.task = task;
+			this.taskControl = task.getControl();
+
 		}
 
 		@Override
 		public void run() {
-			if (aborted) {
-				return;
-			}
-
-			if (Thread.currentThread().isInterrupted()) {
-				return;
-			}
+			CloseableIteration<T, QueryEvaluationException> res = null;
 
-			ParallelExecutor<T> taskControl = task.getControl();
+			try {
 
-			if (taskControl.isFinished()) {
-				return;
-			}
+				if (aborted || Thread.currentThread().isInterrupted() || taskControl.isFinished()) {
+					throw new InterruptedException();
+				}
```

Comment:
I would need to think and look a bit more what the "aborted" state actually means (i.e. if it is also kind of valid state where only partial results are requested from the client). In this case we would need to make sure that the exception is not preventing the consumption of the partial results.

Another question: why would you want to kill the threads entirely? They are controlled by a thread-pool and are re-used for operations. If I remember correctly the thread pool has some time-to-live before it removes threads.

Note: this is currently just out of my head - I do not have a chance to look closer into the code right now (could do in the beginning of next year after the Christmas / New Years time).

For the `isInterrupted()` case it is perfectly fine to propagate the exception

Reply:
I don't know what `isFinished()` or `aborted` mean either. The problem that I see is that `ControlledWorkerUnion` uses a Phaser to wait for all the tasks to complete (I think), and that is why I think we should always call either `toss(...)` or `done()`.

I wonder if there are any existing solutions within Java or some kind of library that would provide a more robust and documented solution to solving these concurrency issues.

Reply:
Also, about the threads. What I'm seeing is that the test sometimes stalls for a while after it's actually done. It seems to be because it's waiting for the thread pools to terminate and at least one thread is stuck waiting on a phaser.

Reply:
Ok, I had another look into the code and tried to remind myself about the implementation. I think it should be fine to throw the InterruptedException also in the `aborted` state: looks like this is some artifact which currently is not used at all. Not sure if it rather should be removed entirely.

What I meant above with consumption of partial results / cancelling queries, is done on another level (i.e. when scheduling the operations for intermediate results)

If you have any alternative to using the phaser for blocking until tasks are completed, I am happy to learn about it.